walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,128 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
-public final class FindMeetingQuery {
+/**
+ * Returns a Collection of times which are available for optional attendees and mandatory attendees.
+ * If optional attendees cannot be fit in, then only returns the times which mandatory attendees can
+ * attend
+ *
+ * @param events a list of events that may need to be accounted for
+ * @param request the specifications for the meeting to be scheduled
+ * @return a Collection of available times for requested meeting
+ */
+public final class FindMeetingQuery implements Comparator<Event> {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<Event> eventList = events.stream().collect(Collectors.toList());
+    Collections.sort(eventList, new FindMeetingQuery());
+    int mandatoryEnd = TimeRange.START_OF_DAY;
+    int optionalEnd = TimeRange.START_OF_DAY;
+    Collection<TimeRange> mandatoryAvailableTimes = new ArrayList<TimeRange>();
+    Collection<TimeRange> optionalAvailableTimes = new ArrayList<TimeRange>();
+    for (int i = 0; i < eventList.size(); i++) {
+      mandatoryEnd = checkIfEventConflicts(
+          mandatoryAvailableTimes, eventList.get(i), request, false, mandatoryEnd);
+      optionalEnd = checkIfEventConflicts(
+          optionalAvailableTimes, eventList.get(i), request, true, optionalEnd);
+    }
+    addFinalRange(mandatoryEnd, request, mandatoryAvailableTimes);
+    addFinalRange(optionalEnd, request, optionalAvailableTimes);
+
+    return (optionalAvailableTimes.isEmpty() && !request.getAttendees().isEmpty())
+        ? mandatoryAvailableTimes
+        : optionalAvailableTimes;
+  }
+
+  /**
+   * Comparator which organizes events based off the earliest start time
+   */
+  public int compare(Event a, Event b) {
+    // sort by start
+    return a.getWhen().start() - b.getWhen().start();
+  }
+
+  /**
+   * Checks if the set of event attendees has any members from the request
+   *
+   * @param eventAttendees the set of people attending the event
+   * @param request the specifications of the request used to retrieve attendees and optional
+   *     attendees
+   * @param checkingOptional a boolean if we are checking for optional attendees
+   * @return if any of the requested attendees exist in the list for eventAttendees
+   */
+  public static boolean hasAttendees(
```

Comment:
This method is called first by a method further down on the page... please re-org file to have top-down structure (e.g. methods appear in the order in which they're called).

Reply:
Add top down structure